asmcomp/polling.ml

```diff
@@ -0,0 +1,256 @@
+            List.fold_left
+              (fun tmp_rec_handlers (_, handler) ->
+                tmp_rec_handlers @ find_rec_handlers handler)
+              [] handlers
+          in
+          let body_rec_handlers = find_rec_handlers body in
+          body_rec_handlers @ non_rec_catch_handlers @ find_rec_handlers f.next
+      )
+  | Itrywith (body, handler) ->
+      let handler_rec_handler = find_rec_handlers handler in
+      let body_rec_handlers = find_rec_handlers body in
+      body_rec_handlers @ handler_rec_handler @ find_rec_handlers f.next
+  | Iexit _ | Iend | Ireturn
+  | Iop (Itailcall_ind _)
+  | Iop (Itailcall_imm _)
+  | Iraise _ ->
+      []
+  | Iop _ -> find_rec_handlers f.next
+
+(* given the list of handler ids [rec_handelrs] for recursive catches, add polls before
+   backwards edges starting from Mach instruction [i] *)
+let instrument_body_with_polls (rec_handlers : int list) (i : Mach.instruction)
+    =
+  (* the [current_handlers] list allows for an optimisation which avoids putting a poll
+    before the first jump in to a loop *)
+  let rec instrument_body (current_handlers : int list) (f : Mach.instruction) =
+    let instrument_with_handlers i = instrument_body current_handlers i in
+    match f.desc with
+    | Iifthenelse (test, i0, i1) ->
+        {
+          f with
+          desc = Iifthenelse (test, instrument_with_handlers i0, instrument_with_handlers i1);
+          next = instrument_with_handlers f.next;
+        }
+    | Iswitch (index, cases) ->
+        {
+          f with
+          desc = Iswitch (index, Array.map instrument_with_handlers cases);
+          next = instrument_with_handlers f.next;
+        }
+    | Icatch (rec_flag, handlers, body) ->
+        {
+          f with
+          desc =
+            Icatch
+              ( rec_flag,
+                List.map
+                  (fun (idx, instrs) ->
+                    (idx, instrument_body (idx :: current_handlers) instrs))
+                  handlers,
+                instrument_with_handlers body );
+          next = instrument_with_handlers f.next;
+        }
+    | Itrywith (body, handler) ->
+        {
+          f with
+          desc = Itrywith (instrument_with_handlers body, instrument_with_handlers handler);
+          next = instrument_with_handlers f.next;
+        }
+    | Iexit id ->
+        let new_f = { f with next = instrument_with_handlers f.next } in
+        if List.mem id current_handlers && List.mem id rec_handlers then
+          add_fused_poll_before new_f
+        else new_f
+    | Iend | Ireturn | Iop (Itailcall_ind _) | Iop (Itailcall_imm _)
+      ->
+        f
+    | Iraise _ -> 
+      add_checked_poll_before true { f with next = instrument_with_handlers f.next }
+    | Iop _ -> { f with next = instrument_with_handlers f.next }
+  in
+  instrument_body [] i
+
+(* adding a poll to these functions is rarely what we want *)
+let ignored_functions = [ "caml_apply2"; "caml_apply3" ]
```

Comment:
Why only ignore these two arities, and not higher arity caml_apply calls?
Also, it might be good to justify why other common calls (caml_curry or caml_tuplify) are not ignored.

Reply:
This is something that @mshinwell has pointed out actually, we should be excluding anything with the `caml_apply` prefix and could probably extend to a few others.

The ignored functions are common ones where we know they're safe to not include polls in but where the analysis may not be able to determine that.